### PR TITLE
improve index_select performance on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76869
* __->__ #76868

Improve some hotspot operators in transformer models.

This PR is targeting at further improving `torch.index_select` performance on CPU.

`index_select` is a critical hotspot in decoder-based Transformer variants with TopK sampling approach, also it is widely used in other scenarios like `nn.Embedding`.

The existing implementation on `index_select` has a fast path when `dim == 1`, this PR parallels it.

The existing implementation is generally OK when used in embedding,  e.g. with input like '[50000, 128]', while the index_size (50000) is big enough. it will parallel on index_size.

But for transformer shapes, the current implementation is not very efficient: The transformer shape is like [5, 8, N, 64] and N ranges from 2 to 114 and dim = 2.

* when N is very small: e.g. N=2, the input shape is [5, 8, 2, 64], it is a sequential run, actually most of time is spent is on constructing TensorIterator itself since the problem size is very small.
* when N is median size: e.g. N=50, slice_size < grain_size and 5 * slice_size > grain_size, it will parallel on index_size (e.g. 5). The magic number 5 is topk, it is not very friendly for parallelization: a) if you have 20 cores, 3/4 the CPU will be idle; b) when you have 4 cores, it will go 2-2-1 and still 1 core is idle.
* when N is large: e.g. N = 114, slice_size > grain_size: it will parallel on slice_size. it will launch 5 omp sessions which will increase omp parallel overhead.

The kernels in this PR is explained in the notes, and briefly:
* when outer_size == 1 (dim=0): parallel on index_size; if slice_size is very big, do blocking to increase parallelism;
* when outer_size != 1 (dim !=0): parallel on outer_size; if the inner_size is very small, use vectorized gather instead of scalar logic.

The following is benchmark result from torch operator_benchmark, Xeon 6248, dual sockets, 20 cores per socket, 2.5GHz.
```
numactl --physcpubind=0 --membind=0 python -m pt.index_select_test --omp_num_threads 1 --mkl_num_threads 1 --tag_filter ${tag_name}
```

Unit: us per iteration.

### index_select_test tag "short"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_M8_N8_K1_dim1_cpu | 2.392 | 2.507 | 95.41% | 2.380 | 2.531 | 94.03%
index_select_M256_N512_K1_dim1_cpu | 62.509 | 42.695 | 146.41% | 61.886 | 19.812 | 312.37%
index_select_M512_N512_K1_dim1_cpu | 137.012 | 102.238 | 134.01% | 133.378 | 21.863 | 610.06%
index_select_M8_N8_K2_dim1_cpu | 2.398 | 2.441 | 98.24% | 2.509 | 2.603 | 96.39%
index_select_M256_N512_K2_dim1_cpu | 493.737 | 108.078 | 456.83% | 490.053 | 20.706 | 2366.72%
index_select_M512_N512_K2_dim1_cpu | 982.585 | 197.707 | 496.99% | 1022.278 | 44.885 | 2277.55%

### index_select_test tag "long"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_M128_N128_K1_dim1_cpu | 6.090 | 4.767 | 127.75% | 5.934 | 4.739 | 125.22%
index_select_M128_N128_K2_dim1_cpu | 17.238 | 6.327 | 272.45% | 17.010 | 6.337 | 268.42%
index_select_M128_N1024_K1_dim1_cpu | 30.587 | 20.686 | 147.86% | 30.283 | 11.935 | 253.73%
index_select_M128_N1024_K2_dim1_cpu | 247.052 | 72.236 | 342.01% | 240.516 | 13.607 | 1767.59%
index_select_M1024_N128_K1_dim1_cpu | 25.972 | 16.416 | 158.21% | 25.837 | 9.257 | 279.11%
index_select_M1024_N128_K2_dim1_cpu | 123.278 | 54.931 | 224.42% | 122.815 | 11.373 | 1079.88%
index_select_M1024_N1024_K1_dim1_cpu | 375.664 | 286.461 | 131.14% | 350.838 | 81.757 | 429.12%
index_select_M1024_N1024_K2_dim1_cpu | 1902.877 | 537.663 | 353.92% | 1871.703 | 150.800 | 1241.18%

### transformer shapes

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_[5, 8, 2, 64]_dim0 | 4.355 | 2.926 | 148.84% | 4.329 | 2.802 | 154.50%
index_select_[5, 8, 114, 64]_dim0 | 85.808 | 83.773 | 102.43% | 48.573 | 11.305 | 429.66%

### others

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_[50000, 256]_dim0 | 196.599 | 172.371 | 114.06% | 51.246 | 39.631 | 129.31%
index_select_[32, 1024, 128]_dim1 | 2065.429 | 1916.039 | 107.80% | 2035.999 | 596.284 | 341.45%

(*) for shape [50000, 256], prefetch will be helpful.

### CODE REVIEW

**Comment by mingfeima** [comment]
Improve some hotspot operators in transformer models.

This PR is targeting at further improving `torch.index_select` performance on CPU.

`index_select` is a critical hotspot in decoder-based Transformer variants with TopK sampling approach, also it is widely used in other scenarios like `nn.Embedding`.

The existing implementation on `index_select` has a fast path when `dim == 1`, this PR parallels it.

The existing implementation is generally OK when used in embedding,  e.g. with input like '[50000, 128]', while the index_size (50000) is big enough. it will parallel on index_size.

But for transformer shapes, the current implementation is not very efficient: The transformer shape is like [5, 8, N, 64] and N ranges from 2 to 114 and dim = 2.

* when N is very small: e.g. N=2, the input shape is [5, 8, 2, 64], it is a sequential run, actually most of time is spent is on constructing TensorIterator itself since the problem size is very small.
* when N is median size: e.g. N=50, slice_size < grain_size and 5 * slice_size > grain_size, it will parallel on index_size (e.g. 5). The magic number 5 is topk, it is not very friendly for parallelization: a) if you have 20 cores, 3/4 the CPU will be idle; b) when you have 4 cores, it will go 2-2-1 and still 1 core is idle.
* when N is large: e.g. N = 114, slice_size > grain_size: it will parallel on slice_size. it will launch 5 omp sessions which will increase omp parallel overhead.

The kernels in this PR is explained in the notes, and briefly:
* when outer_size == 1 (dim=0): parallel on index_size; if slice_size is very big, do blocking to increase parallelism;
* when outer_size != 1 (dim !=0): parallel on outer_size; if the inner_size is very small, use vectorized gather instead of scalar logic.


The following is benchmark result from torch operator_benchmark, Xeon 6248, dual sockets, 20 cores per socket, 2.5GHz.
```
numactl --physcpubind=0 --membind=0 python -m pt.index_select_test --omp_num_threads 1 --mkl_num_threads 1 --tag_filter ${tag_name}
```

Unit: us per iteration.

### index_select_test tag "short"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_M8_N8_K1_dim1_cpu | 2.392 | 2.507 | 95.41% | 2.380 | 2.531 | 94.03%
index_select_M256_N512_K1_dim1_cpu | 62.509 | 42.695 | 146.41% | 61.886 | 19.812 | 312.37%
index_select_M512_N512_K1_dim1_cpu | 137.012 | 102.238 | 134.01% | 133.378 | 21.863 | 610.06%
index_select_M8_N8_K2_dim1_cpu | 2.398 | 2.441 | 98.24% | 2.509 | 2.603 | 96.39%
index_select_M256_N512_K2_dim1_cpu | 493.737 | 108.078 | 456.83% | 490.053 | 20.706 | 2366.72%
index_select_M512_N512_K2_dim1_cpu | 982.585 | 197.707 | 496.99% | 1022.278 | 44.885 | 2277.55%


### index_select_test tag "long"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_M128_N128_K1_dim1_cpu | 6.090 | 4.767 | 127.75% | 5.934 | 4.739 | 125.22%
index_select_M128_N128_K2_dim1_cpu | 17.238 | 6.327 | 272.45% | 17.010 | 6.337 | 268.42%
index_select_M128_N1024_K1_dim1_cpu | 30.587 | 20.686 | 147.86% | 30.283 | 11.935 | 253.73%
index_select_M128_N1024_K2_dim1_cpu | 247.052 | 72.236 | 342.01% | 240.516 | 13.607 | 1767.59%
index_select_M1024_N128_K1_dim1_cpu | 25.972 | 16.416 | 158.21% | 25.837 | 9.257 | 279.11%
index_select_M1024_N128_K2_dim1_cpu | 123.278 | 54.931 | 224.42% | 122.815 | 11.373 | 1079.88%
index_select_M1024_N1024_K1_dim1_cpu | 375.664 | 286.461 | 131.14% | 350.838 | 81.757 | 429.12%
index_select_M1024_N1024_K2_dim1_cpu | 1902.877 | 537.663 | 353.92% | 1871.703 | 150.800 | 1241.18%


### transformer shapes

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_[5, 8, 2, 64]_dim0 | 4.355 | 2.926 | 148.84% | 4.329 | 2.802 | 154.50%
index_select_[5, 8, 114, 64]_dim0 | 85.808 | 83.773 | 102.43% | 48.573 | 11.305 | 429.66%


### others

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
index_select_[50000, 256]_dim0 | 196.599 | 172.371 | 114.06% | 51.246 | 39.631 | 129.31%
index_select_[32, 1024, 128]_dim1 | 2065.429 | 1916.039 | 107.80% | 2035.999 | 596.284 | 341.45%

(*) for shape [50000, 256], prefetch will be helpful.

**Review by malfet — APPROVED** [review]

**Review by malfet — CHANGES_REQUESTED** [review]
@mingfeima can you please rebase, as import/land attempt fail (build_variables.bzl were moved to top level folder)

**Comment by mingfeima** [comment]
> @mingfeima can you please rebase, as import/land attempt fail (build_variables.bzl were moved to top level folder)

Probably still need to update this one since we recently started the optimization project on PyG. Need to make sure the current optimization is able to cover the scenarios from PyG models...
